src/GraphQL/Language/AST/Fields.cs

```diff
@@ -1,36 +1,91 @@
-using System.Collections;
 using System.Collections.Generic;
+using GraphQL.Execution;
+using GraphQL.Types;
 
 namespace GraphQL.Language.AST
 {
     /// <summary>
     /// Represents a list of field nodes within a document.
     /// </summary>
-    public class Fields : IEnumerable<Field>
+    public class Fields : Dictionary<string, Field>
     {
-        private readonly Dictionary<string, Field> _fields;
-
-        private Fields(Dictionary<string, Field> fields)
+        /// <summary>
+        /// Before execution, the selection set is converted to a grouped field set by calling CollectFields().
+        /// Each entry in the grouped field set is a list of fields that share a response key (the alias if defined,
+        /// otherwise the field name). This ensures all fields with the same response key included via referenced
+        /// fragments are executed at the same time.
+        /// <br/><br/>
+        /// See http://spec.graphql.org/June2018/#sec-Field-Collection and http://spec.graphql.org/June2018/#CollectFields()
+        /// </summary>
+        public Fields CollectFrom(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet)
         {
-            _fields = fields;
+            List<string> visitedFragmentNames = null;
+            CollectFields(context, specificType, selectionSet, ref visitedFragmentNames);
```

Comment:
I added `ref` to create temporary `List<string>` for tracking fragment names only when needed, i.e. when fragments are used in the request. Previously, this list was created every time before calling the method.